### PR TITLE
scheduler/lock_manager: Handle the corner case that resumable pessimstic lock request is pushed to queue after cancelling

### LINKED ISSUE #13298: Enhanced queueing for conflicting pessimistic transactions
## Background

Currently, we are adopting a simple but rough implementation for pessimistic lock waiting and waking up. It works just fine in most normal cases, however, there will be problems under workloads that has frequent lock conflicts. When there are many requests trying to acquire the lock on a single key in TiKV:

* The order that the transactions successfully get the lock has too much randomness
* When a pessimistic lock request is woken up after waiting a while in TiKV, it always returns `WriteConflict` to TiDB so that TiDB will retry the statement.
* When many pessimistic lock requests are waiting on the same key in TiKV and the current lock is released, one of the requests will be woken up, and the others will be woken up too after delaying for `wake-up-delay-duration` (which is a configurable value that defaults to 10ms). If the first awakened transaction's retrying request arrives late, the key might already have been locked by another transaction.

This may lead to some problems:

* The latency to acquire the lock might be very unstable, causing high tail latency.
* There might be too many statement-retries that are useless but wastes resources.
* When the problem is severe, it's also possible to cause errors like lock wait timeout and statement retry limit exceeded.

By carefully adjusting `wake-up-delay-duration`, we are actually able to try to make it work well in most scenarios. However, it's very hard to adjust it since it's difficult to predict the effect of changing it. It's even harder if a user who don't understand the internal implementation of TiKV want to tune it. Besides, its a store level config, which means, if a user have more than one workloads that has different characteristics in a single cluster, it's impossible to use different `wake-up-delay-duration` for different workloads.

We expects it would be better to make transactions in such kind of scenarios to execute sequentially, avoiding too much useless statement retries and high tail latency. However, this will be very complicated to implement in our current architecture, especially considering that we needs to guarantee the backward compatibility. We have a complicated design for this problem and it took us much efforts to trying to demonstrate it. We've confirmed that the complicated design is helpful for solving the problem stated above in those kinds of scenarios. However, the stronger queueing behavior in lock waiting introduces higher average latency, or to say, lower QPS. We do have ideas to optimize it further, however, it would make the design much more complicated.

Now we are going to productize the idea. As the first step, we will first implement a subset of the original design: we will enhance the queueing behavior for pessimistic lock requests that involves only one key. This will be helpful for many cases that has frequent lock conflicts on some single rows. And since there's possibly QPS regression, the new behavior will be optional and disabled by default. A user will need to choose to get lower tail latency at the expense of higher average latency if the user needs.

<details>
<summary>Demo branches</summary>

The draft branches for demo can be found at:

* https://github.com/pingcap/tidb/pull/35588
* https://github.com/tikv/client-go
* https://github.com/tikv/tikv/pull/12749

Most part of the final implementation can be ported from these branches, but there will also be some differences.

</details>

## Tasks

- [ ] The RFC / design document
  - RFC for TiKV: https://github.com/tikv/rfcs/pull/100
  - Design doc for TiDB: https://github.com/pingcap/tidb/pull/37518
- [ ] The spec document

- kvproto
    - [x] Protocol change (kvproto) https://github.com/pingcap/kvproto/pull/872
- TiKV
    - [x] Lock waiting queue implementation
        - Each key will have its own waiting queue. In the demo branch, it's put in `Latch`, but later we found hard to store it with an efficient and simple data structure. We will consider store it in the lock table of `ConcurrencyManager` or another separated concurrent hash map.
        - https://github.com/tikv/tikv/pull/13486
        - https://github.com/tikv/tikv/pull/13560
    - [x] Waiter manager & deadlock detector adaption
        - `WaiterManager` no longer have the responsibility to wake up lock requests after the lock is released. Instead, we will need it to be able to index requests by unique tokens. Therefore both its interface and inner data structure need to be updated.
        - https://github.com/tikv/tikv/pull/13447
    - [x] New implementation of storage command `AcquirePessimisticLock`
        - For requests that contains only one key, it can choose to use the new locking behavior: it's allowed to force lock the key even when there's write conflict. It will use the new protocol to return the actual locked ts (the actual lower bound of `for_update_ts`) to the client (TiDB).
        - https://github.com/tikv/tikv/pull/13687
    - [x] Resuming of pessimistic lock requests
        - For pessimistic lock requests that choose to use the new locking behavior, after waiting for another lock, we allow it to continue acquiring the lock after being woken up, instead of returning `WriteConflict` to the client to request a retry.
        - https://github.com/tikv/tikv/pull/13833
        - https://github.com/tikv/tikv/pull/13826
    - [x] Lock wait info update for queueing requests
        - When many requests are queued waiting the same key, and one of them is granted the lock after the previous lock being released, then for the other queueing requests, the transaction they are waiting for is changed. In this case, we need some update operations so that it can perform correct deadlock detection and provide diagnostic information about lock waiting.
- client-go
    - [x] Support aggressive locking behavior
      - https://github.com/tikv/client-go/pull/528
- TiDB
    - [x] Support aggressive locking behavior which can be enabled by system variable
      - https://github.com/pingcap/tidb/pull/35588
- Misc tasks and issues
  - [x] https://github.com/tikv/tikv/pull/13865
  - [x] Telemetry
  - [ ] Confirm compatibility with lazy constraint check in TiDB's pessimistic transaction
  - [ ] Find a proper behavior for the case where insert (should_not_exist) constraint check failure and write conflict occurs together https://github.com/tikv/tikv/issues/14293
  - [ ] In some rare corner cases, for_update_ts collision may cause unexpected pessimistic-rollback https://github.com/pingcap/tidb/issues/41792
  - [ ] In some rare corner cases, replayed pessimistic lock request on a key that lost its pessimistic lock may affect transaction correctness https://github.com/tikv/tikv/issues/14311
- Optional optimizations
    - [ ] Region level waking up after region change (leader or epoch change)
    - [ ] Batch re-detect deadlocks after updating lock wait info
    - [ ] Replace the implementation of `Key` from `Vec<u8>` to something like `Arc<Vec<u8>>` or `Arc<[u8]>` to reduce overhead of cloning keys (it needs to be tested whether it helps)
- Integration tests

### PR DESCRIPTION
**Pull request by MyonKeminta**
### What is changed and how it works?

Issue Number: Close #13298

What's Changed:

```commit-message
scheduler/lock_manager: Handle the corner case that resumable pessimstic lock request is pushed to queue after cancelling.

When a lock-waiting request is woken up and continues its execution, in some cases it's possible that it encounters other transaction's lock agian. In this case, the entry will be put to the lock waiting queue again. However, there might be problem when LockManager tries to cancel the request (due to timeout or other possible errors. This PR handles this case.
```

The detailed time-sequence of this corner case is explained in the comments in the code.
There's also some fixes to the previously merged code.

### Related changes

\-

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/storage/txn/scheduler.rs [hunk 1]
```diff
@@ -1011,19 +1008,23 @@ impl<E: Engine, L: LockManager> Scheduler<E, L> {
...
-                            // TODO: Handle not-none case when supporting resumable pessimistic lock
-                            // requests.
-                            assert!(res.is_none());
+                            if let Some(resumable_lock_wait_entry) = res {
+                                self2.schedule_awakened_pessimistic_locks(
+                                    None,
+                                    None,
+                                    smallvec![resumable_lock_wait_entry],
+                                );
+                            }
```
Comment (MyonKeminta) [inline src/storage/txn/scheduler.rs:1027]: This change should have been done in the previous PR but mistakenly missing and not covered by the tests. It's now covered by the new failpoint test.

File: src/storage/lock_manager/lock_wait_context.rs [hunk 2]
```diff
@@ -53,6 +62,108 @@ pub struct LockWaitContextSharedState {
...
+    ///     * This is possible to be caused by delayed-waking up or encountering
+    ///       error when writing a lock-releasing command to the engine.
+    /// 4. (lock_manager) At the same time, `LockManager` tries to cancel the
+    /// request due to timeout. But when    calling `finish_request`, the
+    /// entry cannot be found from the `LockWaitQueues`. So it    believes
+    /// that the entry is already popped out and resumed and does nothing.
+    /// 5. (scheduler) An entry is pushed to the `LockWaitQueues` due to
+    /// encountering lock at step 3. 6. Then the request becomes unable to
+    /// be canceled by timeout or other possible errors. In worst
+    ///    cases, the request may stuck in TiKV forever.
```
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply src/storage/lock_manager/lock_wait_context.rs:89]: cargo fmt automatically formatted it lke this...

File: src/storage/lock_manager/lock_wait_context.rs [hunk 3]
```diff
@@ -53,6 +62,108 @@ pub struct LockWaitContextSharedState {
...
+    ///
+    /// To solve this problem, a `is_canceled` flag should be set when
+    /// `LockManager` tries to cancel it, before accessing the
+    /// `LockWaitQueues`; when an entry is pushed to the `LockWaitQueues`,
+    /// check if `is_canceled` is set after locking its inner map (ensures
+    /// exclusive access with `LockManager`), and if it's set, cancel the
+    /// request like how `LockManager` should have done.
```
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply src/storage/lock_manager/lock_wait_context.rs:96]: I think so. But I didn't find any solution to avoid managing the state in two sperated data structure. At least lock waiting queue and lock manager should be in a single module in the future.

File: src/storage/lock_manager/lock_wait_context.rs [hunk 4]
```diff
@@ -53,6 +62,108 @@ pub struct LockWaitContextSharedState {
...
+
+    pub fn is_canceled(&self) -> bool {
+        self.is_canceled.load(Ordering::Acquire)
+    }
+
+    pub(in crate::storage) fn get_external_error(&self) -> StorageError {
+        self.external_error_rx.lock().recv().unwrap()
+    }
+
+    fn put_external_error(&self, error: StorageError) {
```
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply src/storage/lock_manager/lock_wait_context.rs:156]: Yes.

File: src/storage/lock_manager/lock_wait_context.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sticnarf) [reply src/storage/lock_manager/lock_wait_context.rs:89]: `cargo fmt` doesn't do that at my place. 🤔 It only limits the line length, but it should not merge multiple lines into one.

**Review by sticnarf — APPROVED** [review]
LGTM

File: src/storage/lock_manager/lock_wait_context.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply src/storage/lock_manager/lock_wait_context.rs]: It did this when my lines are too long... and now I've fixed it..

File: src/storage/lock_manager/lock_wait_context.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (cfzjywxk) [reply src/storage/lock_manager/lock_wait_context.rs:99]: Maybe we could think about it in the future.

File: src/storage/lock_manager/lock_wait_context.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (cfzjywxk) [reply src/storage/lock_manager/lock_wait_context.rs:166]: Should we add some assertions?

**Review by cfzjywxk — APPROVED** [review]

File: src/storage/lock_manager/lock_wait_context.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply src/storage/lock_manager/lock_wait_context.rs:171]: Added
